### ISSUE TITLE
class/id attributes

### ISSUE THREAD

**Issue by leeoniya** [markedjs/marked#1431]
a bit of a resurrection of #116.

would it be possible to add support for special attrs?

https://michelf.ca/projects/php-markdown/extra/#spe-attr

they are extremely useful and can solve a lot of things without resorting to html, like text alignment, image floats, image margins, responsive image sizing, alt attrs, title attrs, etc.

it's a smallish feature (i think) with big ux/dx improvements, imo.

thanks!

**Comment by styfle** [comment]
It does sound useful! But this feature doesn't seem to be standardized in the CommonMark spec or the GFM spec.

There's some discussion about this feature in [this thread](https://talk.commonmark.org/t/consistent-attribute-syntax/272).

The solution (as you mentioned) is to use html input and marked will just pass it through to the output.

**Comment by leeoniya** [comment]
> The solution (as you mentioned) is to use html input and marked will just pass it through to the output.

my goal is to make a small internal blogpost editor for non-technical users. i easily can teach them to author markdown (and store markdown in db) but would like to avoid teaching them html and also avoid a whole host of security issues by simply disallowing/filtering out all html.

not having to teach/write/edit/read markup if a major ux and cognitive win, imo. any idea how involved an extension like this is likely to be?

**Comment by styfle** [comment]
> would like to avoid teaching them html and also avoid a whole host of security issues by simply disallowing/filtering out all html

I'm not sure you're going to avoid security issues because attributes need to be [escaped too](https://stackoverflow.com/q/9187946/266535).

You might want to consider that depending who your audience is and if there is arbitrary user input the will end up in attributes.

**Comment by leeoniya** [comment]
you're right, though my intention was never to support arbitrary attrs & values. if the parser only accepts chars that do not need to be escaped, that would be sufficient for 100% of my cases and probably 95% of most cases. `/` might be the one exception but not sure it would matter if `<` is blacklisted. anyways, +1 for it being an extension, then i can tweak it as needed without having to wait for consensus of the commonmark spec committee or whatever.

**Comment by leeoniya** [comment]
ok, after a bit of prototyping, i think i've narrowed down what's generically needed in `marked` core to support my use-case without having to override every renderer.

TL;DR, all of `marked`'s renderers should handle inserting arbitrary `attrs` that may exist on any `token`. this way i dont need to touch all the renderers and simply intercept the tokenizer to strip & parse the attrs in my extension:

```js
const attrsCache = {};

const RE_ID      = /#([-\w]+)/;
const RE_CLASSES = /\.([-\w.]+)/;
const RE_ATTRS   = /\[([^=\]]+)=?([^\]]+)?\]/g;

function parseAttrs(raw) {
  if (attrsCache[raw] == null) {
    let m, attrs = {};

    if (m = raw.match(RE_ID))
      attrs.id = m[1];

    if (m = raw.match(RE_CLASSES))
      attrs.class = m[1].replace(/\./g, " ");

    while (m = RE_ATTRS.exec(raw))
      attrs[m[1]] = m[2];

    let hasAttrs = false;

    for (let attr in attrs) {
      hasAttrs = true;
      break;
    }

    attrsCache[raw] = hasAttrs ? attrs : null;
  }

  return attrsCache[raw];
}

let tokens = marked.lexer('# my {#foo.bar.baz[moo=cow]}');

const RE_ATTRS2 = /^(.+)\s+\{(.+)\}$/;

tokens.forEach(t => {
  if (t.type != 'html') {
    let m = t.text.match(RE_ATTRS2);
    t.text = m[1];
    t.attrs = parseAttrs(m[2]);
  }
});

console.log(tokens);
```

**Comment by leeoniya** [comment]
upon further investigation, it looks like `this.token` is not passed down to the renderer methods. what do you guys think is the best way to pass this info down without polluting the args?

the only thing i can think of with the current architecture is to add a mutating `renderer.token` that's updated by the parser at the top of the switch/case in `Parser.prototype.tok` so it can be accessed as `this.token` from within the renderer methods to assemble the attribute strings.

**Comment by fxck** [comment]
So, was this ever merged?

**Comment by UziTech** [comment]
No, it is not in the commonmark or GFM spec so it most likely won't be a part of marked. We did create a better way to [extend marked](https://marked.js.org/#/USING_PRO.md#use) in v1.0.0 so someone could create an extension to add this functionality.
